cardano-api/test/Test/Cardano/Api/Typed/Value.hs

```diff
@@ -32,14 +33,41 @@ prop_roundtrip_Value_flatten_unflatten =
 -- Remember that Maps cannot have duplicate keys and therefore
 -- we will never go from Value -> ValueNestedRep (via toValueNestedRep) to a
 -- ValueNestedRep with duplicate values.
+
 prop_roundtrip_Value_unflatten_flatten :: Property
 prop_roundtrip_Value_unflatten_flatten =
     property $ do
       v <- forAll genValueNestedRep
-      let v' = valueToNestedRep (valueFromNestedRep v)
-      v `equiv` v'
-  where
-    equiv (ValueNestedRep a) (ValueNestedRep b) = sort a === sort b
+      canonicalise v === valueToNestedRep (valueFromNestedRep v)
+
+canonicalise :: ValueNestedRep -> ValueNestedRep
+canonicalise (ValueNestedRep bundles) =
+  ValueNestedRep . filter removeZeroAda $ mergeDuplicates bundles
+ where
+  mergeDuplicates :: [ValueNestedBundle] -> [ValueNestedBundle]
+  mergeDuplicates l =
+    let allAssets :: [(PolicyId, Map AssetName Quantity)]
+        allAssets = [(pId, qMap) | (ValueNestedBundle pId qMap) <- l]
+
+        allAda :: [Quantity]
+        allAda = [ada | ValueNestedBundleAda ada <- l]
+
+        summedMinted :: Map PolicyId (Map AssetName Quantity)
+        summedMinted = Map.fromListWith (Map.unionWith (<>)) allAssets
+
+        summedAda :: Quantity
+        summedAda = sum allAda
+
+        flattenedMinted = map (uncurry ValueNestedBundle) $ Map.toList summedMinted
+        flattenedAda = ValueNestedBundleAda summedAda
+
+    in sort $ flattenedAda : flattenedMinted
+
+  removeZeroAda :: ValueNestedBundle -> Bool
+  removeZeroAda (ValueNestedBundleAda 0) = False
+  removeZeroAda (ValueNestedBundle _ aNameQmap) =
+    0 `notElem` Map.elems aNameQmap
+  removeZeroAda _ = True
```

Comment:
This is a filter predicate not a filter itself, so call it something like `filter (not . isZeroAda)`

But also, filtering at this level cannot work. We need to filter out the zero quantities in the nested bundles, and omit the ada if it itself is zero.

If the test did not catch this, it shows we don't have well-tuned generators. I think we should adjust the `Value` and `ValueNestedRep` so 1 out of 10 times it samples from the full range, and 9 out of 10 times from a very small (e.g. 5) range of policy ids (A..E), asset names (A..E) and quantities (-2..2). That will mean we frequently get asset names with the same policy id, and quantities that sum to 0.